sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -307,6 +309,32 @@ object RemoveRedundantProject extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Remove [[Sort]] in subqueries that do not affect the set of rows produced, only their
+ * order. Subqueries produce unordered sets of rows so sorting their output is unnecessary.
+ */
+object RemoveSubquerySorts extends Rule[LogicalPlan] {
+
+  /**
+   * Removes all [[Sort]] operators from a plan that are accessible from the root operator via
+   * 0 or more [[Project]], [[Filter]] or [[View]] operators.
+   */
+  private def removeTopLevelSorts(plan: LogicalPlan): LogicalPlan = {
+    plan match {
+      case Sort(_, _, child) => removeTopLevelSorts(child)
+      case Project(fields, child) => Project(fields, removeTopLevelSorts(child))
+      case Filter(condition, child) => Filter(condition, removeTopLevelSorts(child))
+      case View(tbl, output, child) => View(tbl, output, removeTopLevelSorts(child))
+      case _ => plan
+    }
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Subquery(child) => Subquery(removeTopLevelSorts(child))
+    case SubqueryAlias(name, child) => SubqueryAlias(name, removeTopLevelSorts(child))
```

Comment:
`SubqueryAlias` is not the subquery you want. This is just an alias of a query/table/view.  For example, 
```Scala
Seq((1, 2, "1"), (3, 4, "3")).toDF("int", "int2", "str_sort").orderBy('int.asc).as('df1)
```

Reply:
Thanks! I've been trying to understand the role of `Subquery` and `SubqueryAlias`. My confusion is that subqueries do seem to get planned as `SubqueryAlias` operators, e.g.:

    scala> spark.sql("SELECT count(*) from (SELECT id FROM dft ORDER BY id)").explain(true)
    == Parsed Logical Plan ==
    'Project [unresolvedalias('count(1), None)]
    +- 'SubqueryAlias __auto_generated_subquery_name
       +- 'Sort ['id ASC NULLS FIRST], true
          +- 'Project ['id]
             +- 'UnresolvedRelation `dft`

In the example you give I (personally) think it's still reasonable to drop the ordering, but understand that might surprise some users. It wouldn't be hard to skip the root if it's a subquery - but what do you propose for detecting subqueries if my method isn't right?

Reply:
```Seq((1, 2, "1"), (3, 4, "3")).toDF("int", "int2", "str_sort").orderBy('int.asc).as('df1)```

Before entering optimizer, we get rid of `SubqueryAlias` by the rule `EliminateSubqueryAliases`. Basically, it is no-op after query analysis. The name is a little bit confusing, I have to admit.

Reply:
Yep, that's why I added the new rule just before `EliminateSubqueryAliases` (which runs in the optimizer, as part of the 'finish analysis' batch). After `EliminateSubqueryAliases` there doesn't seem to be any way to detect subqueries.

Another approach I suppose would be to handle this like `SparkPlan`'s `requiredChildOrdering` - if a parent doesn't require any ordering of the child, (and the child is a `Sort` node), the child `Sort` should be dropped. That seems like a more fundamental change though.